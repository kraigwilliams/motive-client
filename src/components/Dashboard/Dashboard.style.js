import styled from 'styled-components'
import { colors } from '../constants'

export const PageWrapper = styled.div`
  background-color: ${colors.darkgrey};
  margin: 0;
  padding: 0;
  min-height: 100vh;
`;

export const DBHeader = styled.h1`
  text-align: center;
  padding: 30px;
  color: ${colors.white};
  background-color: ${colors.darkgrey};
  line-height: 1;
  border-bottom: .5px solid ${colors.teal};
  border-top: .5px solid ${colors.teal};
  font-weight: bold;
  font-style: italic;
  font-size: 35px;
`;

export const ContentWrapper = styled.div`
  display: flex;
  flex-direction: column;
  background-color: ${colors.darkgrey};
  padding: 30px;
  margin: 0;
  height: 100vh;
  align-content: center;

  @media (min-width: 1000px) {
    flex-direction: row;
    justify-content: initial;
  } 
`;

export const Section = styled.section`
  text-align: center;
  margin: auto;

  .fade-appear {
    opacity: .01;
    transition: opacity 1s ease;
  }

  .fade-appear-active {
    opacity: 1;
    transition: opacity .5s ease;
  }

  @media (min-width: 1000px) {
    margin-top: 0;
  } 

  @media (min-width: 900px) {
    width: 60%;
  }  

  @media (min-width: 500px) {
    width: 80%;
  } 
`;

export const SectionTitle = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 10px;
  margin: auto;
`;