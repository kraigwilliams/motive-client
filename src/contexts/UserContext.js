import React, { Component } from "react";
import AuthApiService from "../services/auth-api-service";
import TokenService from "../services/token-service";
import IdleService from "../services/idle-service";

const UserContext = React.createContext({
  user: {},
  error: null,
  addedConnection: null,
  setError: () => {},
  clearError: () => {},
  setUser: () => {},
  processLogin: () => {},
  processLogout: () => {},
  setAddedConnection: () => {},
});

export default UserContext;

export class UserProvider extends Component {
  constructor(props) {
    super(props);
    this.state = {
      user: {},
      error: null,
      addedConnection: null,
    };

    const jwtPayload = TokenService.parseAuthToken();

    if (jwtPayload) {
      this.state.user = {
        id: jwtPayload.user_id,
        firstname: jwtPayload.firstname,
        lastname: jwtPayload.lastname,
        username: jwtPayload.sub,
      };

      IdleService.setIdleCallback(this.logoutBecauseIdle);
    }
  }

  componentDidMount() {
    if (TokenService.hasAuthToken()) {
      IdleService.registerIdleTimerResets();
      TokenService.queueCallbackBeforeExpiry(() => {
        this.fetchRefreshToken();
      });
    }
  }

  componentWillUnmount() {
    IdleService.unRegisterIdleResets();
    TokenService.clearCallbackBeforeExpiry();
  }

  setError = (error) => {
    this.setState({ error });
  };

  clearError = () => {
    this.setState({ error: null });
  };

  setUser = (user) => {
    this.setState({ user });
  };

  processLogin = (authToken) => {
    TokenService.saveAuthToken(authToken);
    const jwtPayload = TokenService.parseAuthToken();
    this.setUser({
      id: jwtPayload.user_id,
      firstname: jwtPayload.firstname,
      lastname: jwtPayload.lastname,
      username: jwtPayload.sub,
    });
    IdleService.registerIdleTimerResets();
    TokenService.queueCallbackBeforeExpiry(() => {
      this.fetchRefreshToken();
    });
  };

  processLogout = () => {
    TokenService.clearAuthToken();
    TokenService.clearCallbackBeforeExpiry();
    IdleService.unRegisterIdleResets();
    this.setUser({});
  };

  logoutBecauseIdle = () => {
    TokenService.clearAuthToken();
    TokenService.clearCallbackBeforeExpiry();
    IdleService.unRegisterIdleResets();
    this.setUser({ idle: true });
  };

  fetchRefreshToken = () => {
    AuthApiService.refreshToken()
      .then((res) => {
        TokenService.saveAuthToken(res.authToken);
        TokenService.queueCallbackBeforeExpiry(() => {
          this.fetchRefreshToken();
        });
      })
      .catch((err) => {
        this.setError(err);
      });
  };

  setAddedConnection = (value) => {
    this.setState({
      addedConnection: value,
    });
  };

  render() {
    const value = {
      user: this.state.user,
      error: this.state.error,
      addedConnection: this.state.addedConnection,
      setError: this.setError,
      clearError: this.clearError,
      setUser: this.setUser,
      processLogin: this.processLogin,
      processLogout: this.processLogout,
      setAddedConnection: this.setAddedConnection,
    };
    return (
      <UserContext.Provider value={value}>
        {this.props.children}
      </UserContext.Provider>
    );
  }
}
