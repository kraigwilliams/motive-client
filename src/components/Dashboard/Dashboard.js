import React, { Component } from 'react'
import UserContext from '../../contexts/UserContext'
import config from '../../config.js'
import TokenService from '../../services/token-service'
import { colors, PageWrapper } from '../constants'
import CondensedTopic from '../CondensedTopic/CondensedTopic';
import CondensedThought from '../CondensedThought/CondensedThought'

import {DBHeader, Section, ContentWrapper } from './Dashboard.style'


class Dashboard extends Component {
  static contextType = UserContext;


  render() {

    return (
      <PageWrapper>
        <header>
          <DBHeader>
          {this.context.user.name}'s  Motive
          </DBHeader>
        </header>
        <ContentWrapper>
          <Section>
            <h2 style={{color: colors.white}}>
              Topics
            </h2>
            <CondensedTopic 
              title='Coronavirus'
              count='3'
            />
          </Section>
          <Section>
            <h2 style={{color: colors.white}}>
              Thoughts
            </h2>
            <CondensedThought 
              title='Meaning of Life'
            />
          </Section>
        </ContentWrapper>
      </PageWrapper>
    );
  }
}

export default Dashboard