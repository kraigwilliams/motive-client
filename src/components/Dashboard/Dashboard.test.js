import React from "react";
import ReactDOM from "react-dom";
import { shallow, configure } from "enzyme";
import { BrowserRouter as Router } from "react-router-dom";

import toJson from "enzyme-to-json";
import Dashboard from "./Dashboard";
import Adapter from "enzyme-adapter-react-16";
import UserContext from "../../contexts/UserContext";

configure({ adapter: new Adapter() });
describe(`<Dashboard />`, () => {
  //Smoke Testing
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <Router>
        <UserContext.Provider value={{ user: { username: "admin" } }}>
          <Dashboard />
        </UserContext.Provider>
      </Router>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  //Snapshot Testing
  it("renders Dashboard by default", () => {
    const wrapper = shallow(
      <UserContext.Provider>
        <Dashboard />
      </UserContext.Provider>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
