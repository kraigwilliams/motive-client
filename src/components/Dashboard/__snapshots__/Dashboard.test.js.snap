// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> renders Dashboard by default 1`] = `<Dashboard />`;
